### PR TITLE
Improve memory footprint of some test

### PR DESCRIPTION
**Pull request by coot**
# Description

Replaced implementation of `prop_shrink_nonempty` with `prop_shrinkCarefully` since the latter can shrink.
Improved memory footprint of `prop_shrink_nonempty`, e.g. from
```
1,825,176,616 bytes maximum residency (105015 sample(s))
```
to
```
1,147,232 bytes maximum residency (4062 sample(s))
```
as measured with on the `shrink nonequal GovernorMockEnvironment` test.

- ouroboros-network-testing: ShrinkCarefully
- ouroboros-network-testing: improved prop_shrink_nonequal
- peer-selection-test: nightly prop_shrink_nonequal_GovernorMockEnvironment

# Checklist

- Branch
    - [x] Updated changelog files.
    - [x] Commit sequence broadly makes sense
    - [x] Commits have useful messages
    - [x] The documentation has been properly updated
    - [ ] New tests are added if needed and existing tests are updated
    - [ ] If serialization changes, user-facing consequences (e.g. replay from genesis) are confirmed to be intentional.
- Pull Request
    - [x] Self-reviewed the diff
    - [x] Useful pull request description at least containing the following information:
      - What does this PR change?
      - Why these changes were needed?
      - How does this affect downstream repositories and/or end-users?
      - Which ticket does this PR close (if any)? If it does, is it [linked](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue)?
    - [x] Reviewer requested

### CODE REVIEW

**Comment by coot** [comment]
Related to #4687.

**Comment by coot** [comment]
Improved `ouroboros-network-framework:sim-tests` (mostly `Server2` tests) from:

image

to

image

There are two more tests that can be improved:
* `prop_connection_manager_counters`
* `prop_timeouts_enforced`

**Comment by coot** [comment]
The improvement in input-output-hk/typed-protocols#43 is quite significant:
from:
image
to
image

File: ouroboros-network-framework/sim-tests/Test/Ouroboros/Network/Server2/Sim.hs [hunk 1]
```diff
@@ -1206,6 +1206,8 @@ prop_connection_manager_valid_transition_order_racy serverAcc (ArbDataFlow dataF
 -- a better way of injecting god's view traces in a way that the timing issues
 -- aren't an issue.
 --
+-- TODO: this function should be rewritten to analyse the trace as a stream, to
+-- reduce its memory footprint.
```
Comment (bolt12) [inline ouroboros-network-framework/sim-tests/Test/Ouroboros/Network/Server2/Sim.hs:1208]: Should we add an issue for this and track it as good for new-comers?

**Review by bolt12 — COMMENTED** [review]
LGTM, unfortunate how much of the memory increases just by having a call to `counterexample` and the test not failing

**Comment by coot** [comment]
> unfortunate how much of the memory increases just by having a call to counterexample and the test not failing

This is unavoidable. It would be cool to have an option to rerun the failed tests with a `debug` option to capture the trace.

**Review by bolt12 — APPROVED** [review]

File: ouroboros-network-framework/sim-tests/Test/Ouroboros/Network/Server2/Sim.hs [hunk 2]
```diff
@@ -599,13 +598,14 @@ instance Show addr => Show (Name addr) where
     show  MainServer   = "main-server"
 
 
-data ExperimentError addr =
-      NodeNotRunningException addr
-    | NoActiveConnection addr addr
+data ExperimentError =
+      forall addr. (Typeable addr, Show addr) => NodeNotRunningException addr
+    -- | forall addr. (Typeable addr, Show addr) => NoActiveConnection addr addr
```
Comment (bolt12) [inline ouroboros-network-framework/sim-tests/Test/Ouroboros/Network/Server2/Sim.hs:603]: Is this needed?

**Review by bolt12 — APPROVED** [review]

File: ouroboros-network-framework/sim-tests/Test/Ouroboros/Network/Server2/Sim.hs (hunk 2 above)
Comment (coot) [reply ouroboros-network-framework/sim-tests/Test/Ouroboros/Network/Server2/Sim.hs:601]: I left it, so we can come back to it at some point.

**Pull request merged**